### PR TITLE
IPv4 Support

### PR DESCRIPTION
**Pull request by tgraf**
This reworks the handling of the node's address and the corresponding
IPAM interactions and enables allocation of IPv4 addresses. All
existing plugins are updated to make use of the new functionality.

Signed-off-by: Thomas Graf thomas@cilium.io

### CODE REVIEW

File: common/addressing/ip.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	return ip
+}
+
+// Returns the node ID portion of the address or 0
+func (ip CiliumIPv6) NodeID() uint32 {
+	return binary.BigEndian.Uint32(ip[10:14])
+}
+
+// Returns the container ID portion of the address or 0
+func (ip CiliumIPv6) ContainerID() uint16 {
```
Comment (aanm) [inline common/addressing/ip.go]: `EndpointID` instead of `ContainerID` maybe?

Comment (tgraf) [inline common/addressing/ip.go]: Good point. I'll change it.

**Comment by tgraf** [comment]
Ping @aanm: Please review the go bits

File: bpf/ctmap/ctmap.go [hunk 2]
```diff
@@ -62,124 +134,120 @@ func (m *CtMap) Dump() (string, error) {
...
-		if err != nil {
-			break
-		}
-
-		err = bpf.LookupElement(
-			m.Fd,
-			unsafe.Pointer(&nextKey),
-			unsafe.Pointer(&entry),
-		)
+func (m *CtMap) doGc(interval uint16, key unsafe.Pointer, nextKey unsafe.Pointer) bool {
```
Comment (aanm) [inline bpf/ctmap/ctmap.go]: Is there a case where the return value is `true`?

File: bpf/ctmap/ctmap.go [hunk 3]
```diff
@@ -62,124 +134,120 @@ func (m *CtMap) Dump() (string, error) {
...
 
-func OpenMap(path string) (*CtMap, error) {
-	fd, _, err := bpf.OpenOrCreateMap(
-		path,
-		C.BPF_MAP_TYPE_HASH,
-		uint32(unsafe.Sizeof(CtKey{})),
-		uint32(unsafe.Sizeof(CtEntry{})),
-		MAX_KEYS,
-	)
+func (m *CtMap) GC(interval uint16) int {
```
Comment (aanm) [inline bpf/ctmap/ctmap.go:234]: Is there a case where deleted is != 0?

File: bpf/go/map_ctrl.go [hunk 4]
```diff
@@ -285,7 +285,7 @@ func mainBPFUpdateKey(ctx *cli.Context) {
 		return
 	}
 
-	u16key := uint16(key)
+	u16key := uint32(key)
```
Comment (aanm) [inline bpf/go/map_ctrl.go]: Is a `u16key` or a `u32key`?

File: bpf/go/map_ctrl.go [hunk 5]
```diff
@@ -319,7 +319,7 @@ func mainBPFDeleteKey(ctx *cli.Context) {
...
```
Comment (aanm) [inline bpf/go/map_ctrl.go]: Is a `u16key` or a `u32key`?

File: bpf/lxcmap/lxcmap.go [hunk 6]
```diff
@@ -142,16 +142,34 @@ func (m *LXCMap) WriteEndpoint(ep *types.Endpoint) error {
...
 	}
 
-	return bpf.DeleteElement(m.fd, unsafe.Pointer(&id))
+	// FIXME: errors are currently ignored
+	id6 := uint32(ep.ID)
+	err1 := bpf.DeleteElement(m.fd, unsafe.Pointer(&id6))
+
+	if id4 := uint32(ep.IPv4.EndpointID()); id4 != 0 {
+		id4 = id4 | (1 << 16)
+		if err2 := bpf.DeleteElement(m.fd, unsafe.Pointer(&id4)); err2 != nil {
```
Comment (aanm) [inline bpf/lxcmap/lxcmap.go]: If you write `if err1 := bpf.DeleteElement(m.fd, unsafe.Pointer(&id4)); err1 != nil {`, the value in `err1` on line 163 would stay the same.

```
foo := 2
if foo := 3; foo == 3 {
  // foo is 3 here
}
// foo is 2 here
```

File: daemon/daemon/ct.go [hunk 7]
```diff
@@ -8,12 +8,32 @@ import (
...
+	fd, err := bpf.ObjGet(file)
+	if err != nil {
+		log.Warningf("Unable to open CT map %s: %s\n", file, err)
+		return
+	}
+
+	f := os.NewFile(uintptr(fd), file)
+	m := ctmap.CtMap{Fd: fd, Type: ctType}
+
+	deleted := m.GC(uint16(GcInterval))
```
Comment (aanm) [inline daemon/daemon/ct.go:29]: The returned value it's always `0`

**Comment by aanm** [comment]
@tgraf Comments in line

File: bpf/ctmap/ctmap.go (hunk 2 above)
Comment (tgraf) [inline bpf/ctmap/ctmap.go]: Good catch, should return true to continue gc.

File: bpf/ctmap/ctmap.go (hunk 3 above)
Comment (tgraf) [inline bpf/ctmap/ctmap.go:234]: Used this mainly for debugging initially. Readded.

File: bpf/go/map_ctrl.go (hunk 5 above)
Comment (tgraf) [inline bpf/go/map_ctrl.go]: Was extended in size to carry both IPv4 and IPv6 entries. Renamed to u32key

File: bpf/lxcmap/lxcmap.go (hunk 6 above)
Comment (tgraf) [inline bpf/lxcmap/lxcmap.go]: Right. Fixed.

File: daemon/daemon/ct.go (hunk 7 above)
Comment (tgraf) [inline daemon/daemon/ct.go:29]: Fixed above.

**Pull request merged**